pyconfyg/pyconfyg.py

```diff
@@ -0,0 +1,113 @@
+import ast
+import copy
+import functools
+import itertools
+import re
+import traceback
+
+import astunparse
+
+
+def update_ast(tree, overwrite, allow_double_assignation=False, allow_tuple_assignation=False):
+    met_targets = []
+    for node in tree.body:
+        if not isinstance(node, ast.Assign):
+            continue
+        for target in node.targets:
+            if hasattr(target, "id"):
+                assert allow_tuple_assignation or isinstance(target, ast.Name), \
+                    "Tuple assignation is not allowed in config files (e.g. `a,b=1,2`). Impossible to overwrite '{}' of type '{}'".format(target.id, type(target))
+                assert allow_double_assignation or target.id not in met_targets, \
+                    "Double assignation is not allowed in config files. '{}' seems to be assigned twice.".format(target.id)
+                if target.id in overwrite:
+                    node.value = ast.parse(repr(overwrite.pop(target.id))).body[0].value
+                    met_targets.append(target.id)
+    # Add remaining keys
+    for key, value in overwrite.items():
+        tree.body.append(ast.Assign([ast.Name(id=key, ctx=ast.Store())], ast.Constant(value, kind=None)))
+    ast.fix_missing_locations(tree)
+    return overwrite
+
+def product_kwargs(**kwargs):
+    kvs = []
+    for k in kwargs:
+        try:
+            kvs.append([(k, v) for v in kwargs[k]])
+        except BaseException as e:
+            raise SyntaxError(f"Error parsing: `{k}:{kwargs[k]}`. Expected an iterable") from e
+    yield from [dict(kv) for kv in itertools.product(*kvs)]
+
+def parse_strings(*strings, env=None):
+    """ Parses strings of 'k=v' by using builtin exec and returns a dictionary
+        of the created symbols and their value.
+        Arguments:
+            strings - strings to be parsed
+            env     - local environment to use
+    """
+    env = env or {}
+    for i, string in enumerate(strings): # pylint: disable=unused-variable
+        #assert re.match(r"\w*=[\w\[\],'\"\(\)]*", string), \
+        #    f"Failed parsing argument #{i}: \"{string}\". Only \"key=value\" are supported"
+        _exec(string, None, env, description="parsable strings")
+    return env
+
+class InterpreterError(Exception):
```

Comment:
I would rather not define a custom exception class in the middle of the module, but would create a dedicated `exceptions` module containing all your custom exceptions. You could possibly also create a _base_ exception class for your project such as `class PyConfygException(Exception)` and let all your custom exceptions. be subclasses of it: `class InterpreterError(PyConfygException)`